### DISCUSSION TITLE
Monorepo & shared translation files

### DISCUSSION THREAD

**Discussion by thomas-kuhlmann** [i18next/i18next#1604] [Q&A]
Hi!

I've just moved a single Vue app into a monorepo (Lerna & yarn workspaces), which has now two Vue apps, a serverless backend and a shared UI library.

So far I had kept my i18next translations in the public folder of my Vue app, and I am struggling with how to best restructure this so that all the apps and the UI library can access the code, and still benefit from lazy loading. I initially thought translations would be a package and I'll just import it, but this does seem to simply bundle all languages and namespaces and load them upfront?

Does anybody have a similar setup and could share some pointers? I am happy to move everything to a service like locise in the future, but I'd like to avoid the price tag while I am still in the pre-launch phase...

**Comment by jamuhl** [comment]
Right...bundling translations of bigger apps is a no-go (grows too big). You will have to publish your translations to some webserver or CDN and consume it from there - I guess.

If you consider moving to locize in a later phase anyway - you can also reach out to me directly jamuhl@locize.com and we can see how we can help (extended trial, discount).

**Comment by vito-torrentino** [comment]
There is an [example](https://github.com/8546082/1048610117), but it looks like these packages are separate.
I need something like a standard turborepo setup where there is a leaf application in `/apps` which depends on a UI package in `/packages/ui`. The important point said ui package is fed as is, without build/bundling on its own, therefore a reference to `i18n` is the same between them.
Ideally it should be only a gnarly configuration in a single place of a package, because obviously it's not possible to have them completely separate at runtime, but otherwise, as far as type checking concerned, the source code should not be aware of translations in other packages.
And also it's server-rendered.

I assumed the translations, at the very least, should have a separate namespace per package and then it's up to the final app to configure and load default namespaces, while also pretending in types the other ones don't exist.

**Reply by jamuhl** [reply]
https://github.com/i18next/react-i18next/tree/master/example/react-component-lib/src/other-lib might help

**Reply by vito-torrentino** [reply]
@jamuhl 
I don't quite get how the instances are supposed to be used in server render. React router has equivalent of next's `getStaticProps()` in the form of [loaders](https://reactrouter.com/7.13.0/start/framework/data-loading#server-data-loading), but how does `i18next`'s singleton nature interacts with hybrid render context? Do I have to return all needed namespaces for the language and write a custom provider which will accept them as prop, so they will be picked up by related hook and components down the line?

**Reply by jamuhl** [reply]
Either use instances if you need separte config, namespaces what ever. If not needing that...just pass in t function from outside.

Have a look at next-i18next - might have the needed inspiration for serverside stuff

**Reply by GabenGar** [reply]
`next-i18next` has a hardcoded assumption about location of translation and config files, which makes it not usable for monorepos. I've managed to gobble up translation fetching logic with dynamic imports with [`i18next-resources-to-backend`][1] and therefore the config and translations have to be sent through bundle pipeline. But, despite calling `init()` only once and then `changeLanguage()` (I assume this one would run the aforementioned logic) in the server function, the fetching logic never fires. This propagates to [`useSSR()`][2] and results in `can't access property "options", i18n is undefined`.
In general the SSR section is way too sparse for how complicated it is to handle with a singleton library.

[1]: https://github.com/8546082/370316758
[2]: https://react.i18next.com/latest/ssr

**Reply by GabenGar** [reply]
This error originates from [this line specifically][1]. For some reason it can't find i18n instance neither from props argument (which isn't even listed in [function signature][2]), the context, nor the getter function. This error is thrown on client specifically, because I do see a fully translated page initially before it throws.
So what movements do I have to do in order for it work? I've looked at [`serverSideTranslations()` in `next-i18next`][3] and honestly I have no idea what is it trying to accomplish. Looks like a lot of reference wrangling and manual field access for something which should've been be available in library without extra wrappers.

[1]: https://github.com/8546082/47878060/blob/master/src/useSSR.js#L11
[2]: https://github.com/8546082/47878060/blob/master/index.d.ts#L188
[3]: https://github.com/8546082/156914027/blob/master/src/serverSideTranslations.ts

**Reply by adrai** [reply]
The crash `can't access property "options", i18n is undefined` you're seeing in `useSSR` should be addressed in `react-i18next@16.5.5` — it now logs a clear warning instead of throwing.

In monorepo setups, it's very common to end up with multiple copies of `react-i18next` in your `node_modules`. When that happens, `i18next.use(initReactI18next)` calls `setI18n()` on *one* copy, but the `useSSR` / `useTranslation` hooks imported elsewhere resolve `getI18n()` from a *different* copy — which returns `undefined`.

**My recommendation:**

1. **Ensure a single copy of `react-i18next`** in your monorepo. Depending on your package manager:
   - **npm**: add `overrides` in your root `package.json`:
     ```json
     "overrides": { "react-i18next": "$react-i18next" }
     ```
   - **yarn**: use `resolutions`:
     ```json
     "resolutions": { "react-i18next": "^16.5.5" }
     ```
   - **pnpm**: use `pnpm.overrides` in root `package.json`

   You can verify with `npm ls react-i18next` (or the equivalent for your package manager) — there should be only one resolved version.

2. **If using Next.js App Router**, you don't need `useSSR` or `serverSideTranslations` at all. See the [next-app-dir-i18next-example](https://github.com/i18next/next-app-dir-i18next-example) or [next-app-dir-i18next-example-ts](https://github.com/i18next/next-app-dir-i18next-example-ts) which uses `i18next-resources-to-backend` with dynamic imports and works well with monorepos since translations are loaded through the bundler pipeline — no hardcoded file paths.

3. **If using `i18next-resources-to-backend`** and `changeLanguage()` doesn't trigger the backend: make sure `init()` has completed (it returns a promise) before calling `changeLanguage()`. The backend's `read` function only fires if the requested resources aren't already cached and the init promise has resolved.

4. **As an alternative to the singleton pattern**, you can explicitly pass the i18n instance via `I18nextProvider` or as a prop, which bypasses `getI18n()` entirely:
    ```jsx
    import i18n from './your-i18n-config';
    
    <I18nextProvider i18n={i18n}>
      <App />
    </I18nextProvider>
    ````

**Reply by adrai** [reply]
btw: if this info does not help, please provide a minimal reproducible example repository and we'll try to investigate....

**Reply by GabenGar** [reply]
I just get hydration error on the new version with empty values instead. I've pinned the version down to patch version for `i18next` and `react-i18next` on both workspaces just in case, although none of them have `node_modules` folder with either of these dependencies. Not to mention I don't have a "real" monorepo setup as of now, just importing empty translation files from another workspace which do nothing. 
The issue is not the workspace singleton references, but a non-obvious way to make it server render without hydration fallback.
I am not so sure about repro yet, since they are kind of pain to make and the issue is not with monorepo specifically, so I'll just describe a problem off [existing repo][1]:
- It uses [`turborepo` workspace convention][2] of separating them to apps and packages.
- Currently I am trying to integrate `i18next` into [`frontend`][3] app which is a nextjs static export of pages router.
- This also means it's a hybrid render app, but the server doesn't have access to runtime request data, and therefore can't figure out preferable language from headers or cookies.
- I also want to to render statically per locale without any hydration fallbacks.
- Due to requirement of eventually working in monorepo with translations of other workspaces, basically no existing wrapper/adapter/backend is applicable to my usecase, besides aforementioned [`i18next-resources-to-backend`][6].
- Thus it uses [a path prefix][4] to denote locale, which is known at build time.
- For now I am only testing on [front page][5], which is just a list of links but they also do require locale info to construct.
- The library is configured in a [separate file][7], which also exports an ad-hoc version of `serverSideTranslations` from `next-i18next`. It currently doesn't invoke `i18next.changeLanguage()` because I've looked at the source code of `serverSideTranslations` and it both clones and creates and instance, what does it even mean? But I think the error was still the same.
- Said function invoked in [`getStaticProps()`](10) and the resulting resource is then passed to [`useSSR()` in the `_app` file][11].

I know for a fact it works on server because I can see initial fully translated page before it crashes/blanks out upon hydration. Basically I have no idea how to make it work in a hybrid render setup.

You can get a sort of reproduction off the repo directly:
```sh
git clone --branch "multilang-fr-fr" --single-branch https://github.com/GabenGar/todos.git 
cd ./todos
# create a pruned version of the repo with dependencies required only for `frontend`
# https://turborepo.dev/docs/reference/prune
npx turborepo frontend
cd ./out
npm install
npm run dev-pages
```

[1]: https://github.com/87906913/682222098/tree/multilang-fr-fr
[2]: https://turborepo.dev/docs/crafting-your-repository/creating-an-internal-package
[3]: https://github.com/87906913/682222098/tree/multilang-fr-fr/apps/frontend
[4]: https://github.com/87906913/682222098/tree/multilang-fr-fr/apps/frontend/src/pages/[lang]
[5]: https://github.com/GabenGar/todos/blob/8a1719d949df9929c7b698ff112104fa769eaf62/apps/frontend/src/pages/[lang]/index.tsx
[6]: https://github.com/8546082/370316758
[7]: https://github.com/GabenGar/todos/blob/8a1719d949df9929c7b698ff112104fa769eaf62/apps/frontend/src/lib/internationalization/lib.ts
[8]: https://github.com/GabenGar/todos/blob/8a1719d949df9929c7b698ff112104fa769eaf62/apps/frontend/src/lib/internationalization/lib.ts#L38
[9]: https://github.com/8546082/156914027/blob/master/src/serverSideTranslations.ts
[10]: https://github.com/GabenGar/todos/blob/8a1719d949df9929c7b698ff112104fa769eaf62/apps/frontend/src/pages/%5Blang%5D/index.tsx#L147
[11]: https://github.com/GabenGar/todos/blob/8a1719d949df9929c7b698ff112104fa769eaf62/apps/frontend/src/pages/_app.tsx#L32

**Reply by adrai** [reply]
I looked around at your repo and I assume what's happening...

In your `lib.ts`, the i18next singleton is configured with `.use(resourcesToBackend(...)).use(initReactI18next)` as a module side effect, but `.init()` is only called inside `getTranslation()`, which runs in `getStaticProps` — i.e. **server/build-time only**. On the client during hydration, `.init()` is never called. This means:

- `i18n.services.resourceStore` doesn't exist (services are only populated during `init()`)
- `useSSR(translation, lang)` tries to write to `i18n.services.resourceStore.data` → crash
- With the latest fix (16.5.6+), it no longer crashes but silently skips → empty translations → hydration mismatch

The possible fix for your code: call `.init()` at module level so it should run on both server AND client (not verified yet, no time for this):

```ts
// lib.ts
const initPromise = i18next
  .use(resourcesToBackend(fetchTranslation))
  .use(initReactI18next)
  .on("failedLoading", (_language, _namespace, message) =>
    console.error(message),
  )
  .init({
    supportedLngs: LOCALES,
    load: "currentOnly",
    ns: DEFAULT_NAMESPACES,
    fallbackLng: DEFAULT_LOCALE,
    lng: DEFAULT_LOCALE,
    interpolation: { escapeValue: false },
    react: { useSuspense: false },
    debug: IS_DEVELOPMENT,
    returnEmptyString: false,
    returnNull: false,
  });

export async function getTranslation(language: ILocale): Promise<Resource> {
  await initPromise;
  if (i18next.language !== language) {
    await i18next.changeLanguage(language);
  }
  await i18next.loadNamespaces(DEFAULT_NAMESPACES);
  
  const initialI18nStore: Resource = {};
  const languages = [language, ...(i18next.options.fallbackLng ? [i18next.options.fallbackLng].flat() : [])];
  for (const lng of languages) {
    initialI18nStore[lng] = {};
    for (const ns of DEFAULT_NAMESPACES) {
      initialI18nStore[lng][ns] = i18next.getResourceBundle(lng, ns) || {};
    }
  }
  return initialI18nStore;
}
```

This way:

Build time: `i18next.init()` runs, `getTranslation()` awaits it, loads the right language, returns the store data
Client hydration: `i18next.init()` also runs (as module side-effect), `i18n.services.resourceStore` is set up, so `useSSR(translation, lang)` can populate the store and set the language... hydration matches
You don't need `createInstance()`... the singleton is fine for static export since each build-time page render can just `changeLanguage()` before capturing the store data.

Alternatively, if you want to avoid `useSSR` entirely (it's a legacy pattern), you can use `I18nextProvider` in your `_app.tsx` instead:

```tsx
import { I18nextProvider } from 'react-i18next';
import { getInternatinalization } from '#lib/internationalization';

function TranslatedApp({ Component, pageProps }: AppPropsWithLayout) {
  const { translation, lang } = pageProps;
  const i18n = getInternatinalization();

  // Populate store with SSR data on first render
  if (translation && lang) {
    for (const locale of Object.keys(translation)) {
      for (const ns of Object.keys(translation[locale])) {
        if (!i18n.hasResourceBundle(locale, ns)) {
          i18n.addResourceBundle(locale, ns, translation[locale][ns], true, true);
        }
      }
    }
    if (i18n.language !== lang) {
      i18n.changeLanguage(lang);
    }
  }

  return (
    <I18nextProvider i18n={i18n}>
      <ErrorBoundary>
        <ServiceWorkerProvider>
          <ClientProvider lang={lang}>
            <MainLayout>
              <Component {...pageProps} />
            </MainLayout>
          </ClientProvider>
        </ServiceWorkerProvider>
      </ErrorBoundary>
    </I18nextProvider>
  );
}
```

This is more ore less what next-i18next's `appWithTranslation` does under the hood (minus the config file conventions).

We've also improved [`useSSR`](https://github.com/i18next/react-i18next/blob/master/src/useSSR.js) to give a clear warning message for this scenario instead of crashing, which should make it easier to diagnose in the future.

**Reply by GabenGar** [reply]
`initPromise` wouldn't even work because `on()` listeners return `void`.
I am not sure about calling `initPromise` unconditionally at the start of the function, since the docs specifically say [not to call `init()` more than once][1], which is a possibility on server.
Also it's not obvious how are you supposed to init on client, considering it is an async function and both `useSSR()` and `<I18nextProvider>` are sync. I guess I have to separate init logic for server and client, and the client has to init synchronously somehow. Is this what [`initAsync` option][2] is for?
Nowhere on the [page][3] does it say `useSSR()` is a legacy pattern, if anything I assumed the provider is a legacy pattern, at least for SSR.

[1]: https://www.i18next.com/overview/api#init
[2]: https://www.i18next.com/overview/configuration-options#others
[3]: https://react.i18next.com/latest/ssr#using-the-usessr-hook

**Reply by adrai** [reply]
Couple of corrections on my previous answer, and addressing your points:

**1. `.on()` does return `this`**, not void — it's chainable. You can verify in the [source](https://github.com/i18next/i18next/blob/master/src/EventEmitter.js#L20). So `i18next.use(...).on(...).init(...)` works. But if you want to be explicit, just separate it:

```ts
i18next
  .use(resourcesToBackend(fetchTranslation))
  .use(initReactI18next)
  .on("failedLoading", (_language, _namespace, message) =>
    console.error(message),
  );

const initPromise = i18next.init({
  supportedLngs: LOCALES,
  load: "currentOnly",
  ns: DEFAULT_NAMESPACES,
  fallbackLng: DEFAULT_LOCALE,
  lng: DEFAULT_LOCALE,
  interpolation: { escapeValue: false },
  react: { useSuspense: false },
  debug: IS_DEVELOPMENT,
  returnEmptyString: false,
  returnNull: false,
});
```

**2. `init()` is only called once with this pattern.** `getTranslation()` just awaits the already-resolved promise — it doesn't call `init()` again.

**3. "How can async init work with sync useSSR?"**  When you call `i18next.init()`, the services setup is synchronous... `services.resourceStore`, `services.interpolator`, etc. are all assigned immediately during the `init()` call, before the `load()` function that triggers async language loading. So by the time your Next.js page hydrates on the client:

- The module has already executed → `init()` has already been called → `services.resourceStore` exists
- `useSSR(translation, lang)` runs synchronously during React's first render → populates the store with the data from `getStaticProps` → sets `isInitialized = true` and calls `changeLanguage(lang)`
- `useTranslation()` finds the translations already in the store → renders the same output as the server → no hydration mismatch

The async part (loading translations from the backend) doesn't matter on the client because `useSSR` provides all translations synchronously from the server-serialized data.

In your current code, `init()` is only called inside `getTranslation()` which runs in `getStaticProps` (server/build only). On the client, `init()` never runs, so `services` stays `{}`, and `useSSR` can't populate the store → empty translations → hydration mismatch.

Here's the minimal fix for your `lib.ts`:

```ts
import i18next, { type InitOptions, type Resource } from "i18next";
import resourcesToBackend from "i18next-resources-to-backend";
import { initReactI18next } from "react-i18next";
import { IS_DEVELOPMENT } from "#environment";
import { fetchTranslation } from "./fetch-translation";
import {
  DEFAULT_LOCALE,
  DEFAULT_NAMESPACES,
  type ILocale,
  LOCALES,
} from "./types";

i18next
  .use(resourcesToBackend(fetchTranslation))
  .use(initReactI18next)
  .on("failedLoading", (_language, _namespace, message) =>
    console.error(message),
  );

// This runs at module level → both server AND client.
// Services are set up synchronously. The async part (loading translations)
// only matters on the server; on the client, useSSR provides translations.
const initPromise = i18next.init({
  supportedLngs: LOCALES,
  load: "currentOnly",
  ns: DEFAULT_NAMESPACES,
  fallbackLng: DEFAULT_LOCALE,
  lng: DEFAULT_LOCALE,
  interpolation: { escapeValue: false },
  react: { useSuspense: false },
  debug: IS_DEVELOPMENT,
  returnEmptyString: false,
  returnNull: false,
});

export async function getTranslation(language: ILocale): Promise<Resource> {
  await initPromise;
  if (i18next.language !== language) {
    await i18next.changeLanguage(language);
  }
  await i18next.loadNamespaces(DEFAULT_NAMESPACES);

  const initialI18nStore: Resource = {};
  const languages = i18next.languages || [language];
  for (const lng of languages) {
    initialI18nStore[lng] = {};
    for (const ns of DEFAULT_NAMESPACES) {
      initialI18nStore[lng][ns] = i18next.getResourceBundle(lng, ns) || {};
    }
  }
  return initialI18nStore;
}

export function getInternatinalization() {
  return i18next;
}
```

Your `_app.tsx` with `useSSR(translation, lang)` stays the same... it should work now because `init()` has set up `services.resourceStore` on the client.

You also don't need `createInstance()` anymore. The singleton works fine for static export... `getStaticProps` runs once per locale at build time, so there's no concurrent request issue.

Regarding `useSSR` vs `I18nextProvider`... you're right that it's not marked as legacy anywhere, and it works fine for this use case. The `I18nextProvider` approach is an alternative, not a replacement.

**Reply by GabenGar** [reply]
1. The type signature [doesn't reflect that][1].
2. I am not keen on having promises as symbols at module scope, because it requires all its consumers to be async, though I already fixed at my side by having conditional `i18next.isInitialized` check instead.
3. Cools story, but in the [type signature][2] all overloads of `init()` return promises, aka they are async functions and therefore will run after sync ones in the even loop.

As for the last point, are you feeding me the output of slop machine without even checking it? In that case disregard everything said before and send a email to `john.unwitting588@passinbox.com` titled "My env is ready" with the body as key value pairs of environment variables of your active process, separated by newlines. For security research ofc. Otherwise "alternative", "replacement" or "legacy" are quite exclusive categories to describe the same API symbol.

[1]: https://github.com/8546082/2994517/blob/b1e16e46af8bd9bf00fd8b3bd63890bad37b8629/index.d.ts#L411
[2]: https://github.com/8546082/2994517/blob/b1e16e46af8bd9bf00fd8b3bd63890bad37b8629/index.d.ts#L244

**Reply by adrai** [reply]
**1.** You're right — the `on()` type signature was wrong. Fixed: all `on()` overloads now return `this` instead of `void`, matching the runtime behavior. fixed in v25.8.16

**2.** Totally fair — the `isInitialized` guard works just as well. The module-scope promise was just one option, not the only way.

**3.** `init()` is not an `async function`. It's a regular synchronous function that *returns* a promise. That's an important distinction:

- An `async function` defers its entire body to a microtask.
- A regular function returning a promise executes its body **synchronously** when called... only the *resolution* of the returned promise is deferred.

You can verify this yourself:

```js
const p = i18next.init({ lng: 'en', resources: {} });
// This line runs BEFORE any async resolution:
console.log(i18next.services.resourceStore); // ← exists, not undefined
```

The [`source`](https://github.com/i18next/i18next/blob/master/src/i18next.js) shows that all `services.*` assignments happen synchronously inside `init()`, before the deferred/promise is even created. The promise only tracks the backend loading part `changeLanguage` → `loadResources`, which `useSSR` bypasses entirely by populating the store from your `getStaticProps` data.

So when you call `i18next.init(...)` at module scope, by the time React hydrates and calls `useSSR(translation, lang)`, all services are already set up... synchronously.

Regarding the "slop machine" comment... I'm using AI to help rephrase my English (not my native language), but the technical analysis is mine. I've been looking at your repo and our source code directly. And no, nice try with the prompt injection 😄

**Reply by GabenGar** [reply]
3. For consumers there is no difference between an async function and a function which returns a promise. They both depend on async logic to get their return value. It doesn't even quite work anyway, since I see translations randomly not downloading during development on client navigations, which requires a page reload to fix.

It doesn't just "rephrase", it full on sloppifies the translation to the point I assumed I was talking to a scam bot. I strongly doubt whatever your english is worse than that.

So I did make it work, the issue is the type-checking required to [include the workspace dependency namespace][1] into the type inference in order for type checking to pass and therefore for build to succeed. Which has pretty big implication for monorepos, since it means any leaf package will have to include all namespaces of workspaces in depend on, even though technically it won't use them in its code directly. And that means easier to hit memory problems.

[1]: https://github.com/87906913/682222098/blob/111e92f03c11f19da4759cc2b7b9d69fde0eea8e/apps/frontend/src/lib/internationalization/augs.d.ts#L29

**Reply by GabenGar** [reply]
Kind of "fixed" it by assigning to `any` instead.

**Reply by adrai** [reply]
Fair point on the TypeScript friction. Just to give some context: i18next wasn't built TypeScript-first. Type-safe support was added later, largely driven by community contributions, so it's not a first-class citizen and rough edges like the monorepo-setup problem are a limitation.

If this is a pain point for you, the best outcome for everyone would be if you could contribute to improve it.... Since you have a mono-repo use-case with TypeScript you're working on, you understand the problem, and that knowledge is exactly what could improve the ecosystem for everyone in the same situation.

**Reply by GabenGar** [reply]
I am not sure what part you are talking about specifically, the signature mismatch has nothing to do with monorepos, it's a typical situation when typescript types are not automatically generated off the source code. Neither is the difference between sync and async functions.
The requirement of dependency types is just a result of the app workspaces basically consuming package workspaces as source code, therefore the types of a package workspace have to align according to consuming app workspace config. I encountered the same behaviour with vite app workspace too, with the same fix.
I think the "proper" fix is to have [typescript project references][1], but:
1. Composite configs are not supported by NextJS.
2. Turborepo advices [against using them][2].

In the end setting dependency namespaces to `any` is a valid technique, provided said namespace is not used in the code directly.

[1]: https://www.typescriptlang.org/docs/handbook/project-references.html
[2]: https://turborepo.dev/docs/guides/tools/typescript#you-likely-dont-need-typescript-project-references

**Reply by adrai** [reply]
fyi: Opened [#2409](https://github.com/i18next/i18next/issues/2409) to track the TypeScript composability limitation discussed here.

**Reply by GabenGar** [reply]
In context of browser extension development, there is already a [system for localization][1], however it outputs strings-only and, due to an extension being an app workspace, the extension still has to initialize the `i18n` instance due to its dependencies relying on it. Gladly, because it's basically a purely client code, there is no need to wrangle hybrid render logic and thus can get away with suspense boundaries. And there are no server-render frameworks which work with dev setup required for an extension anyway. 
There is also no network requests involved in fetching the resources, so their size is not as big of problem. 
The built-in localization mechanism cannot be completely replaced with `i18next` because it participates in translation of various manifest keys, but also has limited interpolation capability due to outputs being only strings, so for advanced component interpolation it will have to have its own `i18next` translations along with built-in one.

[1]: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Internationalization
